homeassistant/components/meater/sensor.py

```diff
@@ -0,0 +1,188 @@
+import logging
+
+import async_timeout
+
+from homeassistant.const import DEVICE_CLASS_TEMPERATURE, TEMP_CELSIUS
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.device_registry as dr
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Meater Temperature Probe sensor."""
+    return True
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the entry."""
+    # assuming API object stored here by __init__.py
+    api = hass.data[DOMAIN][entry.entry_id]
+
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                devices = await api.get_all_devices()
+        except Exception as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+        # Populate the entities
+        entities = []
+
+        for dev in devices:
+            if dev.id not in hass.data[DOMAIN]["entities"]:
+                entities.append(
+                    MeaterProbeTemperature(
+                        coordinator, dev.id, TemperatureMeasurement.Internal
+                    )
+                )
+                entities.append(
+                    MeaterProbeTemperature(
+                        coordinator, dev.id, TemperatureMeasurement.Ambient
+                    )
+                )
+                device_registry = await dr.async_get_registry(hass)
+                device_registry.async_get_or_create(
+                    config_entry_id=entry.entry_id,
+                    identifiers={(DOMAIN, dev.id)},
+                    name="Meater Probe " + dev.id,
+                    manufacturer="Apption Labs",
+                    model="Meater Probe",
+                )
+                hass.data[DOMAIN]["entities"][dev.id] = None
+
+        async_add_entities(entities)
+
+        return devices
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="meater_api",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=timedelta(seconds=30),
+    )
+
+    def null_callback():
+        return
+
+    # Add a subscriber to the coordinator that doesn't actually do anything, just so that it still updates when all probes are switched off
+    coordinator.async_add_listener(null_callback)
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+
+
+class MeaterProbeTemperature(Entity):
```

Comment:
Since we are using `DataUpdateCoordinator`, I believe you could inherit from `CoordinatorEntity` here, which provides some built-in methods that you could use.  You should be able to remove the `should_poll`, `async_added_to_hass`, and `async_update` methods as those are included in `CoordinatorEntity`.

https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

Reply:
You are once again correct! Totally missed that in my initial look through the docs. I'll hopefully have some time to update this tonight.

Reply:
Given that a change and it all seems to be working as expected. Good job on spotting that!